OpenDreamRuntime/Objects/DreamList.cs

```diff
@@ -60,32 +65,43 @@ public class DreamList : DreamObject {
 
         public virtual DreamValue GetValue(DreamValue key) {
             if (key.TryGetValueAsInteger(out int keyInteger)) {
-                return _values[keyInteger - 1]; //1-indexed
+                return GetValue(keyInteger - 1); //1-indexed
             }
             if (_associativeValues == null)
                 return DreamValue.Null;
 
             return _associativeValues.TryGetValue(key, out DreamValue value) ? value : DreamValue.Null;
         }
 
+        public virtual DreamValue GetValue(int key) {
+            if (_values == null) throw new IndexOutOfRangeException();
```

Comment:
This PR adds more additional checks making the code more complex and increasing the room for bugs in the future. But, conversely, the optimization it brings up is very questionable and requires to be profiled (see #629).

I would suggest closing this PR and related issue until we will see any optimization problems in any performance test with profiling.

Reply:
I agree, the topic of whether or not this is actually an optimization has come up a few times and I don't see proof that it is.

Turning `GetValues()` into an enumerable is good though, cuts down on large allocations in a few places. Would be good to get that in another PR.